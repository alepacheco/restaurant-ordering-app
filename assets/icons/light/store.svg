<svg xmlns="http://www.w3.org/2000/svg" width="50" height="46" viewBox="0 0 50 46">
  <defs>
    <style>
      .cls-1 {
        fill: #3b3e4a;
        fill-rule: evenodd;
      }
    </style>
  </defs>
  <path id="store" class="cls-1" d="M1843,976l-11.36-24h-27.28L1793,976h7.01v22h36V976H1843Zm-24-22h4.05l5.13,20H1819V954Zm15.01,22v11h-32V976h32ZM1817,954v20h-9.18l5.13-20H1817Zm-20.84,20,9.47-20h5.25l-5.13,20h-9.59Zm5.85,22v-7h32v7h-32Zm28.24-22-5.13-20h5.25l9.47,20h-9.59Z" transform="translate(-1793 -952)"/>
</svg>
