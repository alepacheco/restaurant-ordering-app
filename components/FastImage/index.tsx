import React, { useState, useEffect } from 'react';
import * as FileSystem from 'expo-file-system';
import { Image } from 'react-native';

const cyrb53 = function(str, seed = 0) {
  let h1 = 0xdeadbeef ^ seed,
    h2 = 0x41c6ce57 ^ seed;
  for (let i = 0, ch; i < str.length; i++) {
    ch = str.charCodeAt(i);
    h1 = Math.imul(h1 ^ ch, 2654435761);
    h2 = Math.imul(h2 ^ ch, 1597334677);
  }
  h1 =
    Math.imul(h1 ^ (h1 >>> 16), 2246822507) ^
    Math.imul(h2 ^ (h2 >>> 13), 3266489909);
  h2 =
    Math.imul(h2 ^ (h2 >>> 16), 2246822507) ^
    Math.imul(h1 ^ (h1 >>> 13), 3266489909);
  return 4294967296 * (2097151 & h2) + (h1 >>> 0);
};

const defaultUrl =
  'https://storage.googleapis.com/barapp-data-images/default-dish.jpg';
export const FastImage: React.FC<{ url: string }> = ({ url, ...options }) => {
  const [uri, setUri] = useState(url || defaultUrl);

  useEffect(() => {
    (async () => {
      const { exists } = await FileSystem.getInfoAsync(
        FileSystem.cacheDirectory + cyrb53(url)
      );
      if (exists) {
        setUri(FileSystem.cacheDirectory + cyrb53(url));
      }
    })();
  }, [url]);

  useEffect(() => {
    (async () => {
      const downloadResumable = FileSystem.createDownloadResumable(
        url,
        FileSystem.cacheDirectory + cyrb53(url)
      );
      try {
        const { uri } = await downloadResumable.downloadAsync();

        setUri(uri);
      } catch (e) {
        setUri(defaultUrl);
      }
    })();
  }, [url]);

  return <Image resizeMode="cover" source={{ uri }} {...options} />;
};
