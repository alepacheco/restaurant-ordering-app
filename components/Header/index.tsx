import React from 'react';
import styled from 'styled-components/native';

const Wrapper = styled.SafeAreaView``;

const Title = styled.Text`
  color: ${props => props.theme.textColor};
  font-weight: bold;
  margin: 18px 0 12px 12px;
  font-size: 24px;
`;

export const Header: React.FC<{ title: string }> = ({ title }) => {
  return (
    <Wrapper>
      <Title>{title}</Title>
    </Wrapper>
  );
};
