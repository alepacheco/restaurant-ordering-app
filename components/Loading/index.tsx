import React, { useState, useEffect } from 'react';
import styled from 'styled-components/native';
import LottieView from 'lottie-react-native';

const Wrapper = styled.SafeAreaView<{ background?: string }>`
  display: flex;
  ${props => `background-color: ${props.background || props.theme.color};`}
  height: 100%;
  width: 100%;
  align-items: center;
  margin: auto;
`;

const CenterMiddle = styled.View`
  align-items: center;
  margin: auto;
`;

const LoadingText = styled.Text`
  ${props => `color: ${props.theme.textColor};`}
  text-align: center;
  font-size: 24px;
`;
const StyledLottieView = styled(LottieView)`
  height: 160px;
  width: 160px;
`;

export const Loading: React.FC<{ type?: 'drinks' }> = ({ type }) => {
  const types = {
    default: require('../../assets/animations/loading.json') as string,
    drinks: require('../../assets/animations/drinks-loading.json') as string,
  };

  const selectedType = types[type] || types.default;

  return (
    <Wrapper background={type === 'drinks' ? '#463268' : null}>
      <CenterMiddle>
        <LoadingText>Loading</LoadingText>
        <StyledLottieView
          source={selectedType}
          speed={1}
          loop={true}
          autoPlay
        />
      </CenterMiddle>
    </Wrapper>
  );
};
