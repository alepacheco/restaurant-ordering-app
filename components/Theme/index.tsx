import React from 'react';
import { AppearanceProvider, useColorScheme } from 'react-native-appearance';
import { ThemeProvider } from 'styled-components/native';
import { theme } from './theme';

declare module 'styled-components' {
  export interface DefaultTheme {
    colorScheme: 'dark' | 'light' | 'no-preference';
    color: string;
    textColor: string;

    contrast0_5: string;
    contrast0: string;
    contrast1: string;
    contrast2: string;
    contrast3: string;
    contrast4: string;
    contrast5: string;
    contrast6: string;
    contrast7: string;
    contrast8: string;
    contrast9: string;
    contrast10: string;
  }
}

export const Theme = ({ children }: { children: any }) => {
  return (
    <AppearanceProvider>
      <ThemeProvider
        theme={{
          colorScheme: useColorScheme(),
          ...theme({ colorScheme: useColorScheme() }),
        }}>
        {children}
      </ThemeProvider>
    </AppearanceProvider>
  );
};
