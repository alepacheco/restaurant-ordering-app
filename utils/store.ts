import { createStore, createTypedHooks } from 'easy-peasy';
import model from './models';

const {
  useStoreActions,
  useStoreState,
  useStoreDispatch,
  useStore,
} = createTypedHooks<typeof model>();

export { useStoreActions, useStoreState, useStoreDispatch, useStore };

export const store = createStore(model);

export default store;
