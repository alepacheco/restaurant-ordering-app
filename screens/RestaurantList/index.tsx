import React, { useState, useEffect } from 'react';
import { SafeAreaView, TouchableOpacity, FlatList, Text } from 'react-native';
import { RestaurantEntry } from './RestaurantEntry';
import styled from 'styled-components/native';
import { getLocation, getRestaurants } from './utils';
import { Loading } from './Loading';
import { NoRestaurants } from './NoRestaurants';

const HeaderTitle = styled.Text`
  font-size: 24px;
  margin: 12px;
`;

const HeaderWrapper = styled.View`
  border-bottom-width: 0.5px;
  border-bottom-color: black;
`;

const Separator = styled.View`
  height: 1px;
  width: 100%;
  background-color: rgba(0, 0, 0, 0.2);
`;

const SeparatorWrapper = styled.View`
  display: flex;
  margin: 0 12px;
`;

const ListHeader = ({}) => {
  return (
    <HeaderWrapper>
      <HeaderTitle>Restaurants</HeaderTitle>
    </HeaderWrapper>
  );
};

const OnClickWrapper: React.FC<{ navigate: any; id: number }> = ({
  navigate,
  id,
  children,
}) => {
  return (
    <TouchableOpacity onPress={() => navigate('RestaurantDetails', { id })}>
      {children}
    </TouchableOpacity>
  );
};

export const RestaurantList: React.FC<{ navigation: any }> = ({
  navigation,
}) => {
  const [list, setList] = useState([]);
  const [initialLoad, setInitialLoad] = useState(true);

  const [isLoading, setIsLoading] = useState(true);
  const { navigate } = navigation;

  useEffect(() => {
    if (initialLoad) {
      (async () => {
        try {
          await getRestaurants({ setList, setIsLoading });
        } catch (error) {
          console.warn(error);
        }

        setInitialLoad(false);
      })();
    }

    if (isLoading) {
      (async () => {
        try {
          const location = await getLocation();

          await getRestaurants({ setList, setIsLoading, location });
        } catch (error) {
          console.warn(error);
        }

        setIsLoading(false);
      })();
    }
  }, [isLoading]);

  if (initialLoad) {
    return <Loading />;
  }
  if (list.length === 0 && !isLoading) {
    return <NoRestaurants />;
  }

  return (
    <SafeAreaView>
      <ListHeader />
      <FlatList
        onRefresh={() => setIsLoading(true)}
        refreshing={isLoading}
        data={list}
        renderItem={({ item }) => (
          <OnClickWrapper navigate={navigate} id={item.id}>
            <RestaurantEntry
              title={item.name}
              description={item.description}
              id={item.id}
              distance={item.distance}
            />
          </OnClickWrapper>
        )}
        ItemSeparatorComponent={() => (
          <SeparatorWrapper>
            <Separator />
          </SeparatorWrapper>
        )}
        keyExtractor={item => item.id}
      />
    </SafeAreaView>
  );
};
