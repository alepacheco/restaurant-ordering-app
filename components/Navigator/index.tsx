import React from 'react';
import { createBottomTabNavigator } from 'react-navigation-tabs';
import { Settings } from '../../screens/Settings';
import { createStackNavigator } from 'react-navigation-stack';
import { RestaurantList } from '../../screens/RestaurantList';
import { Profile } from '../../screens/Profile';
import { Search } from '../../screens/Search';
import { RestaurantDetails } from '../../screens/RestaurantDetails';
import { LogIn } from '../../screens/Splash/LogIn';
import { SignUp } from '../../screens/Splash/SignUp';
import { Splash } from '../../screens/Splash';
import { Cart } from '../../screens/Cart';
import { Orders } from '../../screens/Orders';
import { ProductDetails } from '../../screens/ProductDetails';
import { TabBarIcon } from '../TabBar/TabBarIcon';
import { TabBar } from '../TabBar';
import { createAppContainer } from 'react-navigation';

const HomeTabIcon = (props: any) => (
  <TabBarIcon icon="home" focused={props.focused} />
);

const MapTabIcon = (props: any) => (
  <TabBarIcon icon="map" focused={props.focused} />
);

const ProfileTabIcon = (props: any) => (
  <TabBarIcon icon="profile" focused={props.focused} />
);

const BottomTabNavigator = createBottomTabNavigator(
  {
    Restaurants: {
      screen: RestaurantList,
      navigationOptions: {
        title: '',
        tabBarIcon: HomeTabIcon,
      },
    },
    Search: {
      screen: Search,
      navigationOptions: {
        title: '',
        tabBarIcon: MapTabIcon,
      },
    },
    Profile: {
      screen: Profile,
      navigationOptions: {
        title: '',
        tabBarIcon: ProfileTabIcon,
      },
    },
  },
  {
    tabBarComponent: TabBar,
  }
);

const SplashNavigator = createStackNavigator(
  {
    Splash: {
      screen: Splash,
      navigationOptions: {
        headerShown: false,
      },
    },
    LogIn: {
      screen: LogIn,
      navigationOptions: {
        headerShown: false,
      },
    },
    SignUp: {
      screen: SignUp,
      navigationOptions: {
        headerShown: false,
      },
    },
  },
  { initialRouteName: 'Splash' }
);

export const Navigator = createStackNavigator(
  {
    Home: {
      screen: BottomTabNavigator,
      navigationOptions: {
        headerShown: false,
      },
    },
    Settings,

    RestaurantDetails: {
      screen: RestaurantDetails,
      navigationOptions: {
        headerShown: false,
      },
    },
    ProductDetails: {
      screen: ProductDetails,
      navigationOptions: {
        headerShown: false,
      },
    },
    Cart: {
      screen: Cart,
      navigationOptions: {
        headerShown: false,
      },
    },
    Orders: {
      screen: Orders,
      navigationOptions: {
        headerShown: false,
      },
    },
    SplashNavigator: {
      screen: SplashNavigator,
      navigationOptions: {
        headerShown: false,
      },
    },
  },
  { initialRouteName: 'SplashNavigator' }
);

export const Router = createAppContainer(Navigator);
