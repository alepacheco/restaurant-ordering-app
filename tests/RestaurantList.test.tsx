/* eslint-disable @typescript-eslint/ban-ts-ignore */
import React from 'react';
import renderer from 'react-test-renderer';
import axios from 'axios';
import { RestaurantList } from '../screens/RestaurantList';
import { Theme } from '../components/Theme';
import { AppearanceProvider, useColorScheme } from 'react-native-appearance';
import { StoreProvider } from 'easy-peasy';
import { store } from 'store';

jest.useFakeTimers();

jest.mock('axios', () => ({
  get: jest.fn(),
}));
jest.mock('react-native-appearance');
jest.mock('expo-permissions', () => ({
  askAsync: () => ({ status: 'granted' }),
}));
jest.mock('expo-location', () => ({
  getCurrentPositionAsync: () => ({
    coords: {
      longitude: 3,
      latitude: 3,
    },
  }),
}));

useColorScheme.mockImplementation(() => 'light');

// @ts-ignore
AppearanceProvider.mockImplementation(({ children }) => children);

describe('<RestaurantList />', () => {
  it('renders correctly while empty', async () => {
    let tree: null | renderer.ReactTestRenderer = null;

    // @ts-ignore
    axios.get.mockImplementation(() =>
      Promise.resolve({
        data: [],
      })
    );

    await renderer.act(async () => {
      tree = renderer.create(
        <Theme>
          <StoreProvider store={store}>
            <RestaurantList navigation={{ navigate: () => {} }} />
          </StoreProvider>
        </Theme>
      );
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders correctly whith data', async () => {
    let tree: null | renderer.ReactTestRenderer = null;

    // @ts-ignore
    axios.get.mockImplementation(() =>
      Promise.resolve({
        data: [
          {
            _id: '1',
            title: `McDonalds`,
            imageUrl: 'https://picsum.photos/84/84',
            description: 'This is the restaurant description',
          },
          {
            _id: '2',
            title: `McDonalds 2`,
            imageUrl: 'https://picsum.photos/84/84',
            description: 'This is the restaurant description 2',
          },
        ],
      })
    );

    await renderer.act(async () => {
      tree = renderer.create(
        <Theme>
          <StoreProvider store={store}>
            <RestaurantList navigation={{ navigate: () => {} }} />
          </StoreProvider>
        </Theme>
      );
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
