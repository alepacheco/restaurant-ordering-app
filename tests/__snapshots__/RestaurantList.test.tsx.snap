// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RestaurantList /> renders correctly while empty 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <View
    style={
      Array [
        Object {
          "borderBottomColor": "black",
          "borderBottomWidth": 0.5,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 24,
            "marginBottom": 12,
            "marginLeft": 12,
            "marginRight": 12,
            "marginTop": 12,
          },
        ]
      }
    >
      Restaurants
    </Text>
  </View>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Array [
        Object {
          "display": "flex",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 24,
            "textAlign": "center",
          },
        ]
      }
    >
      No Restaurants nearby
    </Text>
  </RCTSafeAreaView>
</RCTSafeAreaView>
`;

exports[`<RestaurantList /> renders correctly whith data 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <View
    style={
      Array [
        Object {
          "borderBottomColor": "black",
          "borderBottomWidth": 0.5,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 24,
            "marginBottom": 12,
            "marginLeft": 12,
            "marginRight": 12,
            "marginTop": 12,
          },
        ]
      }
    >
      Restaurants
    </Text>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "description": "This is the restaurant description",
          "id": "1",
          "image_url": "https://picsum.photos/84/84",
          "title": "McDonalds",
        },
        Object {
          "description": "This is the restaurant description 2",
          "id": "2",
          "image_url": "https://picsum.photos/84/84",
          "title": "McDonalds 2",
        },
      ]
    }
    disableVirtualization={false}
    extraData={
      Object {
        "isScrolling": false,
      }
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Array [
        Object {
          "height": "100%",
        },
        Object {
          "marginBottom": 56,
        },
      ]
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "marginTop": 0,
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "rgba(2, 2, 2, 0.05)",
                    "borderRadius": 8,
                    "display": "flex",
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "marginBottom": 8,
                    "marginLeft": 8,
                    "marginRight": 8,
                    "marginTop": 8,
                    "paddingBottom": 12,
                    "paddingLeft": 12,
                    "paddingRight": 12,
                    "paddingTop": 12,
                  },
                ]
              }
            >
              <Image
                source={
                  Object {
                    "uri": "https://picsum.photos/84/84",
                  }
                }
                style={
                  Array [
                    Object {
                      "borderRadius": 6,
                      "height": 84,
                      "width": 84,
                    },
                  ]
                }
              />
              <View
                style={
                  Array [
                    Object {
                      "display": "flex",
                      "flexBasis": 0,
                      "flexGrow": 2,
                      "flexShrink": 1,
                      "marginLeft": 12,
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "display": "flex",
                        "flexDirection": "row",
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "fontSize": 24,
                          "justifyContent": "center",
                        },
                      ]
                    }
                  />
                  <Text
                    style={
                      Array [
                        Object {
                          "flexBasis": 0,
                          "flexGrow": 1,
                          "flexShrink": 1,
                          "textAlign": "right",
                        },
                      ]
                    }
                  />
                </View>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "gray",
                        "fontSize": 12,
                        "justifyContent": "center",
                      },
                    ]
                  }
                >
                  This is the restaurant description
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "marginTop": 0,
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "rgba(2, 2, 2, 0.05)",
                    "borderRadius": 8,
                    "display": "flex",
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "marginBottom": 8,
                    "marginLeft": 8,
                    "marginRight": 8,
                    "marginTop": 8,
                    "paddingBottom": 12,
                    "paddingLeft": 12,
                    "paddingRight": 12,
                    "paddingTop": 12,
                  },
                ]
              }
            >
              <Image
                source={
                  Object {
                    "uri": "https://picsum.photos/84/84",
                  }
                }
                style={
                  Array [
                    Object {
                      "borderRadius": 6,
                      "height": 84,
                      "width": 84,
                    },
                  ]
                }
              />
              <View
                style={
                  Array [
                    Object {
                      "display": "flex",
                      "flexBasis": 0,
                      "flexGrow": 2,
                      "flexShrink": 1,
                      "marginLeft": 12,
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "display": "flex",
                        "flexDirection": "row",
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "fontSize": 24,
                          "justifyContent": "center",
                        },
                      ]
                    }
                  />
                  <Text
                    style={
                      Array [
                        Object {
                          "flexBasis": 0,
                          "flexGrow": 1,
                          "flexShrink": 1,
                          "textAlign": "right",
                        },
                      ]
                    }
                  />
                </View>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "gray",
                        "fontSize": 12,
                        "justifyContent": "center",
                      },
                    ]
                  }
                >
                  This is the restaurant description 2
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
